### ISSUE TITLE
Private / mangled names in subclasses should not conflict with superclasses

### ISSUE THREAD

**Issue by refi64** [astral-sh/ty#2635]
### Summary

https://play.ty.dev/bf512978-91c0-434c-ab88-1626323b1de8

```python
class X:
    def __init__(self) -> None:
        self.__value = 0
        print('X.value:', self.__get_value())

    def __get_value(self) -> int:
        return self.__value

class Y(X):
    def __init__(self) -> None:
        super().__init__()
        self.__value = 's'
        print('Y.value:', self.__get_value())
    
    def __get_value(self) -> str:  # ERROR: Definition is incompatible with `X.__get_value`
        return self.__value  # ERROR: expected `str`, found `Unknown | Literal["s", 0]`

Y()
```

This isn't actually an error, because the name mangling ensures that both `value` instances are independent:

```
X.value: 0
Y.value: s
```

This works in [pyright](https://pyright-play.net/?code=MYGwhgzhAEAaBcAoaLoBMCmAzaB9XAlgHYEAu%2BAFBBiFgJTQC0AfNAHID2RGSqf01WgDp8ANzAgArhmgBeaAAZk-FAAcATsVIUA5LCHipPHQBoBNLCNwBzDOUPSKdOomUpMOfLfsTHg%2BkysWrwq0Op2kupE5sJivhiuoJAwAJoUsHQh7th4hCTkuFQWDCzsXDxu-BCSqhjqTlbEZJQuoTGWcUZy0DoQOpV8Glq6KQbx8KbtVt64DhhOraiVHrkzc0W0JawQpOpZ-OGkkdH%2BVnOuaXRAA) and [mypy, *partially*](https://mypy-play.net/?mypy=latest&python=3.12&gist=e26b72defe64ae0b8c64b7e5322100e3) (the attribute itself still shows up as an error, seems to be https://github.com/python/mypy/issues/8267).

Note that #645 does exist, but I don't really consider that to be a dupe, since it's primarily about proper name mangling (which is needed for the mentioned external access), which is a bit independent from this, e.g. pyright does *not* allow external access via the mangled names.

### Version

a89bcfa0a

**Comment by carljm** [comment]
Thanks for the report! Yeah, I agree we can track this separately from #645 -- I think we could implement this just by ignoring double-underscore attributes / pretending they don't exist, without actually implementing the name-mangling rules. (Though of course if we did fully implement #645, that would also fix this.)
